### PR TITLE
 Cryptocurrency Formatting and Reputation Scripts

### PR DESCRIPTION
**Pull request by moishce**
## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://github.com/demisto/etc/issues/28894
fixes: demisto/etc/issues/28936

## TDL
- [x] Move Formatting script to the Cryptocurrency pack
- [x] Add a `bitcoin` prefix for validated addresses that were found - Using the formatting script
- [x] Add IOC, and related indicator fields to the pack
- [x] Layout - the default layout with the `Address Type` field in addition.
- [x] Create an integration with `!crypto` command. For now,  this command will set the `Address Type` incident field with `bitcoin` and the ioc reputation with `2` (suspicious)

## Description
Two new scripts to validate a cryptocurrency (support only Bitcoin at the moment) address and enrich it.

### CODE REVIEW

File: Packs/CommonScripts/Scripts/isBitcoin/isBitcoin.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+from hashlib import sha256
+
+DIGITS58 = '123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz'
+
+
+def decode_base58(bitcoin, length):
+    n = 0
+    for char in bitcoin:
+        n = n * 58 + DIGITS58.index(char)
+    return n.to_bytes(length, 'big')
```
Comment (ShahafBenYakir) [inline Packs/CommonScripts/Scripts/isBitcoin/isBitcoin.py:12]: Why not use an existing library for this?

File: Packs/CommonScripts/Scripts/isBitcoin/isBitcoin.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+
+def verify_is_bitcoin(bitcoin):
+    try:
+        bitcoin_bytes = decode_base58(bitcoin, 25)
+        return bitcoin_bytes[-4:] == sha256(sha256(bitcoin_bytes[:-4]).digest()).digest()[:4]
```
Comment (ShahafBenYakir) [inline Packs/CommonScripts/Scripts/isBitcoin/isBitcoin.py:18]: What does this so?

**Review by ShahafBenYakir — APPROVED** [review]

**Review by glicht — DISMISSED** [review]
Script name should start with capital. Let's also call it `Format` as we do with other indicator-format scripts.

Change: `bitcoinAddressValidator` to `BitcoinAddressFormat`

File: Packs/CommonScripts/Scripts/bitcoinAddressValidator/bitcoinAddressValidator.yml [hunk 3]
```diff
@@ -0,0 +1,19 @@
+args:
+- default: false
+  description: The Bitcoin address to verify.
+  isArray: false
```
Comment (glicht) [inline Packs/CommonScripts/Scripts/bitcoinAddressValidator/bitcoinAddressValidator.yml:4]: Need to support `isArray: true`.

Comment (ShahafBenYakir) [reply Packs/CommonScripts/Scripts/bitcoinAddressValidator/bitcoinAddressValidator.yml:4]: @glicht The script should run once for multiple IOCs?

Comment (glicht) [reply Packs/CommonScripts/Scripts/bitcoinAddressValidator/bitcoinAddressValidator.yml:4]: Yes. Take a look at the other format scripts we have for an example.

Comment (ShahafBenYakir) [reply Packs/CommonScripts/Scripts/bitcoinAddressValidator/bitcoinAddressValidator.yml]: @moishce Please validate the when running the script upon multiple indicators, it is filtering only the real bitcoin urls

**Comment by kirbles19** [comment]
@moishce Done.

**Comment by ShahafBenYakir** [comment]
@moishce What is the hold up for this?

**Comment by ShahafBenYakir** [comment]
Can we make sure this pass build?

File: Packs/CommonScripts/Scripts/CryptoCurrenciesFormat/CryptoCurrenciesFormat.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    n = 0
+    for char in address:
+        n = n * 58 + DIGITS58.index(char)
+    return n.to_bytes(length, 'big')
+
+
+def verify_is_bitcoin(address) -> Union[bytes, bool]:
+    try:
+        bitcoin_bytes = decode_base58(address, 25)
+        '''Check if the last four bytes are equal to the
```
Comment (yaakovi) [inline Packs/CommonScripts/Scripts/CryptoCurrenciesFormat/CryptoCurrenciesFormat.py:20]: where this comparison rule came from? please add a link to the source

File: Packs/CommonScripts/Scripts/CryptoCurrenciesFormat/README.md [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+ 
+ 
+## Script Data
+---
+
+| **Name** | **Description** |
+| --- | --- |
+| Script Type | python3 |
+| Tags |  |
+| Demisto Version | 0.0.0 |
```
Comment (yaakovi) [inline Packs/CommonScripts/Scripts/CryptoCurrenciesFormat/README.md:11]: the script has `fromversion`
```suggestion
| Demisto Version | 5.0.0 |
```

**Review by yaakovi — APPROVED** [review]

**Comment by glicht** [comment]
From what I discussed with @michalgold we want to have the reputation as an integration. Also this should be in it's own pack and not the Common.

**Comment by michalgold** [comment]
@glicht yes, discussed with @Bargenish. She'll make the changes in this PR after she is done helping us with a POC blocker.

**Comment by Itay4** [comment]
@Bargenish whats the status here?

**Comment by Bargenish** [comment]
@Itay4 I was assigned to a different issue while working on this and since then the design had changed so this is going to be rewritten after i'll finish with the other issues.

**Review by glicht — COMMENTED** [review]
@michalgold why not in a a separate pack together with the crypto currency indicator?

**Comment by Bargenish** [comment]
@glicht I thought only the new integration should be in a separate pack while the formatting and ioc.
i'll move it from common to the cryptocurrency pack

**Comment by Bargenish** [comment]
@EliorKedar  Please go over the `Packs/Cryptocurrency/IndicatorTypes/reputation-cryptocurrency.json` and `Packs/Cryptocurrency/IndicatorFields/Address_Type.json` files.

**Comment by Bargenish** [comment]
@kirbles19  @ShirleyDenkberg  the Pr has changed and files were added since youv'e reviewed it. Can you please go over the rest of the files?
The CryptoCurrenciesFormat files haven’t  changed.

File: Packs/Cryptocurrency/IndicatorTypes/reputation-cryptocurrency.json [hunk 6]
```diff
@@ -0,0 +1,25 @@
+{
+ "id": "Cryptocurrency Wallet",
```
Comment (EliorKedar) [inline Packs/Cryptocurrency/IndicatorTypes/reputation-cryptocurrency.json:2]: change indicator name to Cryptocurrency Address

File: Packs/Base/Scripts/CommonServerPython/CommonServerPython.py [hunk 7]
```diff
@@ -2285,7 +2288,7 @@ def to_context(self):
 
     class CVE(Indicator):
         """
-        CVE indicator class - https://xsoar.pan.dev/docs/context-standards#cve
+        CVE indicator class - https://xsoar.pan.dev/docs/integrations/context-standards-mandatorycve
```
Comment (bakatzir) [inline Packs/Base/Scripts/CommonServerPython/CommonServerPython.py:2291]:
```suggestion
        CVE indicator class - https://xsoar.pan.dev/docs/integrations/context-standards-mandatory#cve
```

File: Packs/Cryptocurrency/Integrations/Cryptocurrency/Cryptocurrency.yml [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+  required: false
+  type: 0
+- defaultvalue: '50'
+  name: max_fetch
+  required: false
+  type: 0
+- display: Incident type
+  name: incidentType
+  required: false
+  type: 13
```
Comment (bakatzir) [inline Packs/Cryptocurrency/Integrations/Cryptocurrency/Cryptocurrency.yml:22]: by mistake right? we do not have fetch here.

File: Packs/Cryptocurrency/Integrations/Cryptocurrency/Cryptocurrency.yml [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+      type: string
+    - contextPath: Cryptocurrency.Address
+      description: The cryptocurrency address.
+      type: string
+    - contextPath: Cryptocurrency.AddressType
+      description: The cryptocurrency type. e.g. 'bitcoin'.
+      type: string
+    description: Return Cryptocurrency reputation.
+  dockerimage: demisto/python3:3.8.6.13358
+  isfetch: true
```
Comment (bakatzir) [inline Packs/Cryptocurrency/Integrations/Cryptocurrency/Cryptocurrency.yml:62]: rm
```suggestion
```

File: Packs/Cryptocurrency/Integrations/Cryptocurrency/Cryptocurrency.yml (hunk 8 above)
Comment (Bargenish) [reply Packs/Cryptocurrency/Integrations/Cryptocurrency/Cryptocurrency.yml:22]: yes, for some reason its not the latest. we don't have any configuration for this integration. will update this now.

File: Packs/Cryptocurrency/IndicatorFields/Address_Type.json [hunk 10]
```diff
@@ -0,0 +1,30 @@
+{
+ "id": "indicator_addresstype",
```
Comment (EliorKedar) [inline Packs/Cryptocurrency/IndicatorFields/Address_Type.json:2]: Let's change this to Cryptocurrency Address Type

File: Packs/Cryptocurrency/Integrations/Cryptocurrency/Cryptocurrency.py [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+            indicator_type=DBotScoreType.CRYPTOCURRENCY,
+            integration_name=INTEGRATION_NAME,  # Vendor
+            score=score  # Suspicious
+        )
+        crypto_context = Common.Cryptocurrency(
+            address=address,
+            address_type=BITCOIN,
+            dbot_score=dbot_score
+        )
+        hr = f'Cryptocurrency reputation for address {address} was set to {SCORE[score]}'
```
Comment (EliorKedar) [inline Packs/Cryptocurrency/Integrations/Cryptocurrency/Cryptocurrency.py:41]: Let's change the HR to a table with the data, similar to the context output.
Address
AdddressType
Reputation (DBotScore)

Comment (Bargenish) [reply Packs/Cryptocurrency/Integrations/Cryptocurrency/Cryptocurrency.py]: Screen Shot 2020-12-07 at 14 01 37

**Comment by Bargenish** [comment]
@EliorKedar Done

File: Packs/Cryptocurrency/Integrations/Cryptocurrency/Cryptocurrency.py [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+            'Reputation': SCORE[score]
+        }
+        table_name = f'{INTEGRATION_NAME} reputation for {address}'
+        hr = tableToMarkdown(table_name, table_data)
+
+        command_results.append(CommandResults(
+            outputs_prefix='Cryptocurrency',
+            readable_output=hr,
+            outputs_key_field='Address',
+            indicator=crypto_context
```
Comment (yaakovi) [inline Packs/Cryptocurrency/Integrations/Cryptocurrency/Cryptocurrency.py:53]: here and everywhere
https://www.python.org/dev/peps/pep-0008/#when-to-use-trailing-commas
```suggestion
            indicator=crypto_context,
```

File: Packs/Cryptocurrency/Scripts/CryptoCurrenciesFormat/CryptoCurrenciesFormat_test.py [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+import demistomock as demisto
+
+# Bitcoin address, expected_output
+testdata = [
+    ('1AGNa15ZQXAZUgFiqJ2i7Z2DPU2J6hW62i', True),
+    ('1Q1pE5vPGEEMqRcVRMbtBK842Y6Pzo6nK9', True),
+    ('1AGNa15ZQXAZUgFiqJ2i7Z2DPU2J6hW62X', False),
+    ('1ANNa15ZQXAZUgFiqJ2i7Z2DPU2J6hW62i', False),
+    ('1AGNa15ZQXAZUgFiqJ2i7Z2DPU2J6hW62X', False),
+    ('1A Na15ZQXAZUgFiqJ2i7Z2DPU2J6hW62i', False),
```
Comment (yaakovi) [inline Packs/Cryptocurrency/Scripts/CryptoCurrenciesFormat/CryptoCurrenciesFormat_test.py:12]: add a comment why each of them is not a valid bitcoin address

File: Packs/Cryptocurrency/Scripts/CryptoCurrenciesFormat/CryptoCurrenciesFormat.py [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+
+def main():
+    address_list = argToList(demisto.args().get('input'))
+
+    list_results = [f'bitcoin-{address}' for address in address_list if verify_is_bitcoin(address)]
+
+    if list_results:
+        demisto.results(list_results)
+    else:
+        demisto.results("")
```
Comment (yaakovi) [inline Packs/Cryptocurrency/Scripts/CryptoCurrenciesFormat/CryptoCurrenciesFormat.py:36]: consistency over the code
```suggestion
        demisto.results('')
```

**Review by yaakovi — APPROVED** [review]

**Pull request merged**